src/Storages/MergeTree/MergeTreeData.cpp

```diff
@@ -5558,7 +5558,7 @@ std::optional<ProjectionCandidate> MergeTreeData::getQueryProcessingStageWithAgg
     if (settings.parallel_replicas_count > 1 || settings.max_parallel_replicas > 1)
         return std::nullopt;
 
-    auto query_ptr = query_info.original_query;
+    auto query_ptr = query_info.query;
```

Comment:
Is it a requirement to use `original_query` instead of query?

cc: @amosbird

Reply:
It's required. `original_query` is only introduced for this purpose.

Reply:
https://github.com/ClickHouse/ClickHouse/blob/1204f643f1b0f2e8eb92bc844ca8fac3865db3c2/src/Storages/SelectQueryInfo.h#L178

Reply:
I see, do you have a suggestion to work around this problem? There is already similar logic that calls [query.setFinal()](https://github.com/ClickHouse/ClickHouse/pull/40945/files#diff-5452971ae18972aa669265008a29b3732a99d9415f5b3af10d573c0807667a22R519) and I believe the same issue would appear.

Reply:
Projections don't work with FINAL. It seems the FINAL rewrite happens before analysis, so there is no need to change this line.

Reply:
The changes I proposed on this PR do not work if projections are turned on. In projections code, it checks if `query.final()` and then it aborts. The problem is that it checks the original query, not the query that was modified.

Reply:
I see. Then you can tweak the check like this: `if (forceQueryWithFinal() || query.final())`

Reply:
Ok, I'll actually do `if (info.query.final() || query.final())` and add a comment. Same concept applies.

Reply:
It needs to be casted to `ASTSelectQuery`. In order not to mess things up and perform the proper cast, I had to rename the `query` variable to `original_query` and introduce a new one called `query`. I think checking `query.final()` is better than calling `forceQueryWithFinal` because of the following:
1. If for any other reason `query.setFinal()` is called, `forceQueryWithFinal` could return false and the `final` would not be captured. Such case probably already exists, see `needRewriteQueryWithFinal`.
2. The idea behind `forceQueryWithFinal` is to modify a query statement/ object and not be used to control execution flow through the execution.